llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp

```diff
@@ -5833,6 +5834,80 @@ PerformBUILD_VECTORCombine(SDNode *N, TargetLowering::DAGCombinerInfo &DCI) {
   return DAG.getBitcast(VT, PRMT);
 }
 
+static SDValue
+PerformBUILD_VECTOROfTargetLoadCombine(SDNode *N,
+                                       TargetLowering::DAGCombinerInfo &DCI) {
+  // Match: BUILD_VECTOR of v4i8, where first two elements are from a
+  // NVPTXISD::LoadV2 or NVPTXISD::LDUV2 of i8, and the last two elements are
+  // zero constants. Replace with: zext the loaded i16 to i32, and return as a
+  // bitcast to v4i8.
```

Comment:
This seems like a very specific and uncommon occurrence and the test cases look a bit contrived. Is this something we're going to see in real programs?

Reply:
The original use case for this optimization was to clean up after an internal pass. 

However, the test cases in this PR are `llvm-reduce`d IR from real CUDA programs that didn't utilize said internal pass. 

If you think the usefulness isn't worth the maintenance cost, I'm happy to keep it internal.